From akuster@kama-aina.net  Fri Apr 24 15:36:54 2009
From: Armin <akuster@kama-aina.net>
Date: Wed, 22 Apr 2009 12:14:29 -1000
Subject: Sysfs: fix possible memleak in sysfs_follow_link
Subject: bug 13150
To: greg@kroah.com
Message-ID: <49EF96C5.60801@kama-aina.net>

From: Armin Kuster <akuster@mvista.com>

There is the possiblity of a memory leak if a page is allocated and if
sysfs_getlink() fails in the sysfs_follow_link.

Signed-off-by: Armin Kuster <akuster@mvista.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 fs/sysfs/symlink.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/fs/sysfs/symlink.c
+++ b/fs/sysfs/symlink.c
@@ -194,6 +194,9 @@ static void *sysfs_follow_link(struct de
 	unsigned long page = get_zeroed_page(GFP_KERNEL);
 	if (page)
 		error = sysfs_getlink(dentry, (char *) page); 
+		if (error < 0)
+			free_page((unsigned long)page);
+
 	nd_set_link(nd, error ? ERR_PTR(error) : (char *)page);
 	return NULL;
 }
